packages/java/parser-jvm-plugin-backbone/src/main/java/dev/hilla/parser/plugins/backbone/EndpointExposedPlugin.java

```diff
@@ -0,0 +1,168 @@
+        }
+
+        return signature;
+    }
+
+    @Nonnull
+    @Override
+    public NodeDependencies scan(@Nonnull NodeDependencies nodeDependencies) {
+        var node = nodeDependencies.getNode();
+
+        if (node instanceof EndpointNode || node instanceof EndpointExposedNode
+                || node instanceof EndpointNonExposedNode) {
+            // Attach type signature nodes for hierarchy parent superclass and
+            // implemented interfaces, if any
+            var cls = (ClassInfoModel) node.getSource();
+            return nodeDependencies
+                    .appendChildNodes(scanEndpointClassSignature(cls));
+        }
+
+        if (node instanceof EndpointSignatureNode) {
+            // Attach the referenced class from the endpoint hierarchy type
+            // signature
+            var classRef = ((ClassRefSignatureModel) node.getSource());
+            var classInfo = classRef.getClassInfo();
+            return nodeDependencies.appendChildNodes(
+                    Stream.of(createEndpointHierarchyClassNode(classInfo)));
+        }
+
+        return nodeDependencies;
+    }
+
+    /**
+     * Creates and returns nodes for the type signatures of the superclass and
+     * implemented interfaces of the endpoint or endpoint hierarchy ancestor
+     * class.
+     *
+     * <p>
+     * The signatures are used later for type arguments lookup to resolve
+     * generic type variables in exposed method parameter and result types, see
+     * the {@code resolve(Node, NodePath)} lifecycle hook.
+     *
+     * @param endpointClass
+     *            The endpoint or endpoint hierarchy class.
+     * @return The stream of nodes.
+     */
+    private Stream<Node<?, ?>> scanEndpointClassSignature(
+            ClassInfoModel endpointClass) {
+        return Stream
+                .concat(endpointClass.getSuperClass().stream(),
+                        endpointClass.getInterfacesStream())
+                .map(EndpointSignatureNode::of);
+    }
+
+    /**
+     * Creates a node that wraps the given endpoint hierarchy class as a source.
+     * If the class is annotated with the configured {@code @EndpointExposed}
+     * annotation, uses {@code EndpointExposedNode}, otherwise uses
+     * {@code EndpointNonExposedNode}.
+     *
+     * @param classInfo
+     *            The class from the endpoint hierarchy.
+     * @return The node for the class.
+     */
+    private Node<?, ?> createEndpointHierarchyClassNode(
+            ClassInfoModel classInfo) {
+        var endpointExposedAnnotationName = getStorage().getParserConfig()
+                .getEndpointExposedAnnotationName();
+        var exposed = classInfo.getAnnotationsStream()
+                .map(AnnotationInfoModel::getName)
+                .anyMatch(endpointExposedAnnotationName::equals);
+        var classInfoNode = exposed ? EndpointExposedNode.of(classInfo)
+                : EndpointNonExposedNode.of(classInfo);
+        return classInfoNode;
```

Comment:
Yes, but please don't think about performance gains here 🙂 There are even developers that advocate always returning variables as it is easier to read, it helps debugging and has no performance penalties.

Reply:
Actually, it's just Intellij Idea that always recommend me this 😅

Reply:
I would say that I'm not a big fan of supporting debugging introducing that kind of a code. Actually, in my opinion, this change only reduces readability because you expect something to be done with the variable but noting happens. Time to time, it could also become some kind of sacrificing features to the debugging purposes. E.g., Dan Abramov suggests using `return` in arrow functions in JS because it helps readability; but for me, it makes functions to look more complex than they really are.

Reply:
I usually return directly without creating the intermediate variable, but it happens from time to time to have a variable in the code since it was used for debugging. I was just stating that this doesn't make a difference to me. And I fully agree with you regarding returns in arrow functions, by the way.